### PR TITLE
[feature] Type-Aware Field Matcher

### LINKED ISSUE #30: [IDEA] Type-Aware Rule Parsing
I'm curious on your thoughts here. I am using this repo, but mainly just for the rule parsing. I have separate evaluation logic for our use internally. This logic would be improved with the ability to retain the type information from Sigma searches in YAML format. Currently, all values (i.e. float, string, integer) are loaded as strings. If  the value is not scalar, it is empty when loaded by `sigma-go`.

I have tested this with the following modifications to the Field Matcher struct:

```go
type FieldMatcher struct {
	Field     string
	Modifiers []string
	Values    []interface{}
}

func (f *FieldMatcher) unmarshal(field *yaml.Node, values *yaml.Node) error {
	fieldParts := strings.Split(field.Value, "|")
	f.Field, f.Modifiers = fieldParts[0], fieldParts[1:]

	switch values.Kind {
	case yaml.ScalarNode:
		f.Values = []interface{}{nil}
		return values.Decode(&f.Values[0])
	case yaml.SequenceNode:
		return values.Decode(&f.Values)
	case yaml.MappingNode:
		fallthrough
	case yaml.DocumentNode:
		f.Values = []interface{}{map[string]interface{}{}}
		return values.Decode(&f.Values[0])
	case yaml.AliasNode:
		return f.unmarshal(field, values.Alias)
	}
	return nil
}
```
This works as long as you don't import `evaluator`, but breaks the evaluator implementation currently. Since I don't use the evaluator that is bundled here, I wanted to check if this was something you'd be interested in supporting before going any further down this road. To be clear, I'd be okay with making the necessary modifications to the evaluator to support/understand the `interface{}` value types myself, and submit a full PR. I just don't want to go down that path if it's not something you agree with as the project owner.

My use-case is basically that there are some things like `|gt` or `|lt` modifiers which operate on numeric values or more complex custom Sigma additions which require structured/typed data in the field matcher values. For example, internally we have a `|nested` modifier which interprets the value as a separate condition operating on nested documents within the event. This is impossible to implement since the current parser drops any values that aren't scalar.

If you're not interested here, I may end up maintaining an internal fork, but I'd love to push whatever I can back upstream if you're interested. :smiley:

### PR DESCRIPTION
**Pull request by calebstewart**
This PR adds the ability to load arbitrarily  typed values in a field matcher, however maintains the same evaluation logic (i.e. treating all scalar types as strings and ignoring non-scalar typed values). This includes scalar values like integers, strings and floats as well as complex values which lie outside the scope of the Sigma specification, but could be helpful in extensions during parsing or evaluation of rules using this library.

Given your response to the issue I created, it seems you're open to these changes, so I figured I'd just create the PR and work from here. As such, this closes #30 if merged.

### CODE REVIEW

**Comment by calebstewart** [comment]
I think something odd happened, because I was silly and started working on this in my fork off of the branch that was recently merged. As such, the commit history seems to have the commits from the other merge listed. However, the diff looks just fine, so I don't think it's an issue. If the history gets wonky, we can always squash it on merge anyway. :shrug: Sorry about that.

File: evaluator/evaluate_search.go [hunk 1]
```diff
@@ -141,7 +141,24 @@ eventMatcher:
...
+		case int:
+			value = fmt.Sprintf("%v", abstractValue)
+		case float32:
+			value = fmt.Sprintf("%v", abstractValue)
+		case float64:
+			value = fmt.Sprintf("%v", abstractValue)
+		case bool:
+			value = fmt.Sprintf("%v", abstractValue)
+		default:
+			return nil, fmt.Errorf("expected scalar field matching value got: %v", abstractValue)
```
Comment (bradleyjkemp) [inline evaluator/evaluate_search.go:159]: Including the concrete type in the error will help debug this
```suggestion
			return nil, fmt.Errorf("expected scalar field matching value got: %v (%T)", abstractValue, abstractValue)
```

File: rule_parser.go [hunk 2]
```diff
@@ -189,10 +195,17 @@ func (f *FieldMatcher) unmarshal(field *yaml.Node, values *yaml.Node) error {
...
 	case yaml.ScalarNode:
-		f.Values = []string{values.Value}
-
+		f.Values = []interface{}{nil}
+		return values.Decode(&f.Values[0])
 	case yaml.SequenceNode:
 		return values.Decode(&f.Values)
+	case yaml.MappingNode:
+		fallthrough
+	case yaml.DocumentNode:
```
Comment (bradleyjkemp) [inline rule_parser.go:204]: Do we ever expect to see a DocumentNode? My understanding was a DocumentNode was the top-level _thing_ and you couldn't encounter one as some lower down field type

File: evaluator/evaluate_search.go [hunk 3]
```diff
@@ -141,7 +141,24 @@ eventMatcher:
...
+		case string:
+			value = abstractValue
+		case int:
+			value = fmt.Sprintf("%v", abstractValue)
+		case float32:
+			value = fmt.Sprintf("%v", abstractValue)
+		case float64:
+			value = fmt.Sprintf("%v", abstractValue)
+		case bool:
+			value = fmt.Sprintf("%v", abstractValue)
```
Comment (bradleyjkemp) [inline evaluator/evaluate_search.go:157]:
```suggestion
		case int, float32, float64, bool:
			value = fmt.Sprintf("%v", abstractValue)
```

**Review by bradleyjkemp — APPROVED** [review]
Looks good to me!

File: evaluator/evaluate_search.go (hunk 3 above)
Comment (calebstewart) [reply evaluator/evaluate_search.go]: I feel really dumb that I didn't know this was a valid golang syntax. Thank you so much. lol

File: rule_parser.go (hunk 2 above)
Comment (calebstewart) [reply rule_parser.go:204]: That's probably true. I don't have an issue dropping it. I was just going through each potential type, but now that you mention it you're most likely right.

Comment (calebstewart) [reply rule_parser.go]: I'm pretty certain you're right here. I've added a commit to remove that case. :+1:

**Pull request merged**